### ISSUE TITLE
giada 0.16.4 hangs when restarting after setting jack as an audio server

### ISSUE THREAD

**Issue by ycollet** [monocasual/giada#409]
I compiled giada 0.16.6 from sources on Fedora 32 with gcc-10.2.1.
At first start, giada works. I set jack in the config dialog. I close giada and then restart giada and after that, giada always hangs.
If I remove the .giada directory, giada starts again to work until I select jack.
Here is some informations I've got:
```
$ gdb --args /opt/giada-devel/bin/giada 
GNU gdb (GDB) Fedora 9.1-6.fc32
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /opt/giada-devel/bin/giada...
(gdb) r
Starting program: /opt/giada-devel/bin/giada 
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.31-4.fc32.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[init] Giada 0.16.4
[init] Build date: Oct 31 2020
[init] Dependencies:
[init]   FLTK - 1.3.5
[init]   RtAudio - 5.1.0
[init]   RtMidi - 3.0.0
[init]   Libsamplerate
[init]   Libsndfile - libsndfile-1.0.28
[init]   JSON for modern C++ - 3.7.3
[init] Can't read configuration file! Using default values
[midiMapConf::init] scanning midimaps directory '/home/collette/.giada/midimaps/'...
[midiMapConf::init] unable to scan midimaps directory!
[Inferior 1 (process 103505) exited normally]
Missing separate debuginfos, use: dnf debuginfo-install alsa-lib-1.2.3.2-1.fc32.x86_64 bzip2-libs-1.0.8-2.fc32.x86_64 dbus-libs-1.12.20-1.fc32.x86_64 expat-2.2.8-2.fc32.x86_64 flac-libs-1.3.3-2.fc32.x86_64 fltk-1.3.5-3.fc32.x86_64 fontconfig-2.13.92-9.fc32.x86_64 freetype-2.10.4-1.fc32.x86_64 gsm-1.0.18-6.fc32.x86_64 jack-audio-connection-kit-1.9.14-2.fc32.x86_64 libICE-1.0.10-3.fc32.x86_64 libSM-1.2.3-5.fc32.x86_64 libX11-1.6.12-1.fc32.x86_64 libX11-xcb-1.6.12-1.fc32.x86_64 libXau-1.0.9-3.fc32.x86_64 libXcursor-1.2.0-2.fc32.x86_64 libXext-1.3.4-3.fc32.x86_64 libXfixes-5.0.3-11.fc32.x86_64 libXft-2.3.3-3.fc32.x86_64 libXi-1.7.10-3.fc32.x86_64 libXinerama-1.1.4-5.fc32.x86_64 libXrandr-1.5.2-3.fc32.x86_64 libXrender-0.9.10-11.fc32.x86_64 libXtst-1.2.3-11.fc32.x86_64 libasyncns-0.8-18.fc32.x86_64 libcap-2.26-7.fc32.x86_64 libgcc-10.2.1-6.fc32.x86_64 libgcrypt-1.8.5-3.fc32.x86_64 libgpg-error-1.36-3.fc32.x86_64 libpng-1.6.37-3.fc32.x86_64 libsamplerate-0.1.9-5.fc32.x86_64 libsndfile-1.0.28-12.fc32.x86_64 libstdc++-10.2.1-6.fc32.x86_64 libuuid-2.35.2-1.fc32.x86_64 libvorbis-1.3.6-6.fc32.x86_64 libxcb-1.13.1-4.fc32.x86_64 lz4-libs-1.9.1-2.fc32.x86_64 opus-1.3.1-3.fc32.x86_64 pulseaudio-libs-13.99.1-4.fc32.x86_64 rtmidi-3.0.0-6.fc32.x86_64 systemd-libs-245.8-2.fc32.x86_64 xz-libs-5.2.5-1.fc32.x86_64 zlib-1.2.11-21.fc32.x86_64
(gdb) r
Starting program: /opt/giada-devel/bin/giada 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[init] Giada 0.16.4
[init] Build date: Oct 31 2020
[init] Dependencies:
[init]   FLTK - 1.3.5
[init]   RtAudio - 5.1.0
[init]   RtMidi - 3.0.0
[init]   Libsamplerate
[init]   Libsndfile - libsndfile-1.0.28
[init]   JSON for modern C++ - 3.7.3
[midiMapConf::init] scanning midimaps directory '/home/collette/.giada/midimaps/'...
[midiMapConf::init] unable to scan midimaps directory!
[New Thread 0x7fffe9c68700 (LWP 103510)]
[New Thread 0x7fffe9b67700 (LWP 103511)]

Thread 3 "giada" received signal SIG32, Real-time event 32.
[Switching to Thread 0x7fffe9b67700 (LWP 103511)]
0x00007ffff7af79ac in read () from /lib64/libpthread.so.0
(gdb) where
#0  0x00007ffff7af79ac in read () from /lib64/libpthread.so.0
#1  0x00007ffff7c3be32 in Jack::JackClientSocket::Read(void*, int) () from /lib64/libjack.so.0
#2  0x00007ffff7c3f661 in Jack::JackClientNotification::Read(Jack::detail::JackChannelTransactionInterface*) () from /lib64/libjack.so.0
#3  0x00007ffff7c3f496 in Jack::JackSocketClientChannel::Execute() () from /lib64/libjack.so.0
#4  0x00007ffff7c3aba0 in Jack::JackPosixThread::ThreadHandler(void*) () from /lib64/libjack.so.0
#5  0x00007ffff7aed432 in start_thread () from /lib64/libpthread.so.0
#6  0x00007ffff73a0913 in clone () from /lib64/libc.so.6
(gdb)
```

**Comment by tomek-szczesny** [comment]
Hello, and thank you for a fine description of the problem!

However, this gdb output is a bit misleading, as SIG32 is not abnormal in threaded operation. See [this short post](https://gnobal.net/591/what-is-sig32-during-a-gdb-debug-session) on the topic.
We'd be very grateful if you could rerun your debug session ignoring SIG32, to unveil the real problem.
Thanks!

**Comment by ycollet** [comment]
Thanks for the hint.
I produced a real gdb stack now :)
```
Thread 21 "giada" received signal SIG32, Real-time event 32.
[Thread 0x7fffe9b67700 (LWP 104094) exited]
[Thread 0x7fffe9c68700 (LWP 104093) exited]
[New Thread 0x7fffe9c68700 (LWP 104095)]
[New Thread 0x7fffe9b67700 (LWP 104096)]

Thread 1 "giada" received signal SIGSEGV, Segmentation fault.
0x000000000048bd10 in RtApiJack::probeDeviceOpen (this=0x5285a0, device=0, mode=RtApi::OUTPUT, channels=2, firstChannel=4294967294, sampleRate=48000, format=16, 
    bufferSize=0x4d7384 <giada::m::kernelAudio::(anonymous namespace)::realBufsize>, options=0x7fffffffd820) at src/deps/rtaudio/RtAudio.cpp:2318
2318      if ( ports[ firstChannel ] ) {
(gdb) 
(gdb) where
#0  0x000000000048bd10 in RtApiJack::probeDeviceOpen (this=0x5285a0, device=0, mode=RtApi::OUTPUT, channels=2, firstChannel=4294967294, sampleRate=48000, format=16, 
    bufferSize=0x4d7384 <giada::m::kernelAudio::(anonymous namespace)::realBufsize>, options=0x7fffffffd820) at src/deps/rtaudio/RtAudio.cpp:2318
#1  0x0000000000484de8 in RtApi::openStream (this=0x5285a0, oParams=0x7fffffffd7c8, iParams=0x0, format=16, sampleRate=48000, 
    bufferFrames=0x4d7384 <giada::m::kernelAudio::(anonymous namespace)::realBufsize>, callback=
    0x4449b0 <giada::m::mixer::masterPlay(void*, void*, unsigned int, double, unsigned int, void*)>, userData=0x0, options=0x7fffffffd820, errorCallback=0x0)
    at src/deps/rtaudio/RtAudio.cpp:379
#2  0x000000000042e2ff in giada::m::kernelAudio::openDevice () at src/core/kernelAudio.cpp:193
#3  0x0000000000433d02 in giada::m::init::(anonymous namespace)::initAudio_ () at src/core/init.cpp:231
#4  giada::m::init::startup (argc=1, argv=0x7fffffffd9d8) at src/core/init.cpp:231
#5  0x0000000000413637 in main (argc=<optimized out>, argv=<optimized out>) at src/main.cpp:40
```

**Comment by tomek-szczesny** [comment]
> firstChannel=4294967294

Uhum, just as I expected. This is _probably_ a problem that has been already reported and fixed, but oddly enough manifested itself differently than usual.

The problem is related to storing an invalid value of "-1" in certain config file entries. Then, being cast to unsigned integer at some point, you get this.

It is related to #368 and has been already fixed, but there was no release just yet. You might want to check out current master branch straight from github, but be advised this might cause some other problems ;)

To work around your problem, Run AppImage version of Giada to create a valid config file for JACK. Then, after you close Giada, make sure that `"channels_out"` entry in config file is not `-1` - preferably set it to zero.

Let us know if you need further assistance!

**Comment by ycollet** [comment]
I manage to compile giada master branch. And it works nicely.
Just one thing, on fedora, RtMidi.h is installed in /usr/include/rtmidi.
I used the following command line:
```
CFLAGS=-I/usr/include/rtmidi CXXFLAGS=-I/usr/include/rtmidi ./configure --target=linux --prefix=/opt/giada-devel --enable-debug
```

**Comment by tomek-szczesny** [comment]
I don't know much about this, but perhaps this problem arises from the fact that Giada is currently switching to CMake, so the build manual online does not apply anymore. Try using CMake instead and see if it can find a library on its own. The feedback will be appreciated. :)

**Comment by ycollet** [comment]
Excellent ... I am a big cmake fan and I didn't notice the CMakeLists.txt ... Damned !

**Comment by ycollet** [comment]
There is no install target in the cmake file ?

**Comment by tomek-szczesny** [comment]
I don't know anything about that, so I'm calling @monocasual to answer that question :D

**Comment by dromer** [comment]
Yup, also just ran into this. After the first start I changed the settings of audio + midi to jack and after restarting I got:
```
[midiMapConf::init] scanning midimaps directory '/home/dreamer/.giada/midimaps/'...
[midiMapConf::init] unable to scan midimaps directory!
Segmentation fault
```

**Comment by monocasual** [comment]
Good to know the issue appears to be fixed in the development version. Release 0.17.0 will be shipped in a couple of days, please feel free to test it and report back :)

> There is no install target in the cmake file ?

Not yet, and heavily depends on #358

**Comment by hfiguiere** [comment]
> Not yet, and heavily depends on #358

Well, actually, #358 depends on an install target.

**Comment by monocasual** [comment]
This looks like a duplicate of #368, which should have been fixed in 37f0440 .

**Comment by gvnnz** [comment]
Closing as duplicate of #368.
